### PR TITLE
Uploading the repo for Long Range Arena

### PR DESCRIPTION
**Pull request by EvanZhuang**
# Description

> :memo: Uploading the repo for Long Range Arena
>  
> *  Transformer & Linformer are included
> *  A custom task for document matching (AAN) is included  
> * Dataset is hosted on GCP Bucket, see README.md for Details

## Tests

> :memo: Please describe the tests that you ran to verify your changes.
>  
> * Ran Pylint according to [Google Guideline](https://google.github.io/styleguide/pyguide.html) , "Your code has been rated at 9.10/10"
> * Auto formatted with `yapf -i --style='{based_on_style: google, indent_width: 2}' `
> * Ran experiments on Cloud TPU

**Test Configuration**:

## Checklist

- [X] I have signed the [Contributor License Agreement](https://github.com/tensorflow/models/wiki/Contributor-License-Agreements).
- [X] I have read [guidelines for pull request](https://github.com/tensorflow/models/wiki/Submitting-a-pull-request).
- [X] My code follows the [coding guidelines](https://github.com/tensorflow/models/wiki/Coding-guidelines).
- [X] I have performed a self [code review](https://github.com/tensorflow/models/wiki/Code-review) of my own code.
- [X] I have commented my code, particularly in hard-to-understand areas.
- [X] I have made corresponding changes to the documentation.
- [X] My changes generate no new warnings.
- [X] I have added tests that prove my fix is effective or that my feature works.

### CODE REVIEW

File: official/projects/lra/linformer_experiments.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+
+from official.projects.lra import lra_dual_encoder_dataloader
+from official.projects.lra import lra_dual_encoder_task
+
+AdamWeightDecay = optimization.AdamWeightDecayConfig
+PolynomialLr = optimization.PolynomialLrConfig
+PolynomialWarmupConfig = optimization.PolynomialWarmupConfig
+
+
+@dataclasses.dataclass
```
Comment (saberkun) [inline official/projects/lra/linformer_experiments.py:34]: I don't feel we need to subclass.
Could you update according to this example?
https://github.com/tensorflow/models/blob/master/official/nlp/configs/pretraining_experiments.py#L25

File: official/projects/lra/lra_dual_encoder.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    cls_inputs = tf.concat([left_cls_inputs, right_cls_inputs], -1)
+    classifier = layers.ClassificationHead(
+        inner_dim=0 if use_encoder_pooler else cls_inputs.shape[-1],
+        num_classes=num_classes,
+        initializer=initializer,
+        dropout_rate=dropout_rate,
+        name=head_name)
+    predictions = classifier(cls_inputs)
+
+    # b/164516224
```
Comment (saberkun) [inline official/projects/lra/lra_dual_encoder.py:101]: remove?

File: official/projects/lra/lra_dual_encoder.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    # Once we've created the network using the Functional API, we call
+    # super().__init__ as though we were invoking the Functional API Model
+    # constructor, resulting in this object having all the properties of a model
+    # created using the Functional API. Once super().__init__ is called, we
+    # can assign attributes to `self` - note that all `self` assignments are
+    # below this line.
```
Comment (saberkun) [inline official/projects/lra/lra_dual_encoder.py:107]: remove

File: official/projects/lra/lra_dual_encoder_task.py [hunk 4]
```diff
@@ -0,0 +1,319 @@
+# Copyright 2022 The TensorFlow Authors. All Rights Reserved.
```
Comment (saberkun) [inline official/projects/lra/lra_dual_encoder_task.py:1]: Could we add a unit test to the lra dual encoder task?
Test what's the behavior.
I am wondering if we can subclass the nlp/tasks/dual_encoder.py?

File: official/projects/lra/linformer_encoder_block.py [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
+        set to `True` if `norm_first` is `False`.
+      return_attention_scores: If `True`, the output of this layer will be a
+        tuple and additionally contain the attention scores in the shape of
+        `[batch_size, num_attention_heads, seq_dim, seq_dim]`.
+      **kwargs: keyword arguments.
+    """
+    util.filter_kwargs(kwargs)
+    super().__init__(**kwargs)
+
+    # Deprecation warning.
```
Comment (saberkun) [inline official/projects/lra/linformer_encoder_block.py:137]: Remove the support for the deprecated args?

File: official/projects/lra/linformer_encoder_block.py [hunk 6]
```diff
@@ -0,0 +1,446 @@
...
+      target_tensor = input_tensor
+
+    if key_value is None:
+      key_value = input_tensor
+
+    ## Low Rank Projection Here
+    key = self._key_projection(key_value)
+    value = self._value_projection(input_tensor)
+    ## Low Rank Projection Done
+    ## WARNING: ATTENTION MASK
```
Comment (saberkun) [inline official/projects/lra/linformer_encoder_block.py:399]: Remove the warning?

Is the attention mask treatment implemented?

File: official/projects/lra/linformer_encoder_block.py [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+        **common_kwargs)
+    self._value_projection = tf.keras.layers.Dense(
+        self._low_rank_features,
+        activation=None,
+        use_bias=False,
+        kernel_initializer=tf_utils.clone_initializer(self._kernel_initializer),
+        bias_initializer=tf_utils.clone_initializer(self._bias_initializer),
+        name="value_low_rank_projection",
+        **common_kwargs)
+    self._attention_layer = tf.keras.layers.MultiHeadAttention(
```
Comment (saberkun) [inline official/projects/lra/linformer_encoder_block.py:222]: it is still using the MHA?
May you tell the difference of lin former?

File: official/projects/lra/transformer_experiments.py [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+from official.projects.lra import lra_dual_encoder_task
+from official.projects.lra.transformer import TransformerEncoderConfig
+
+AdamWeightDecay = optimization.AdamWeightDecayConfig
+PolynomialLr = optimization.PolynomialLrConfig
+PolynomialWarmupConfig = optimization.PolynomialWarmupConfig
+
+
+@dataclasses.dataclass
+class TransformerOptimizationConfig(optimization.OptimizationConfig):
```
Comment (saberkun) [inline official/projects/lra/transformer_experiments.py:35]: See: https://github.com/tensorflow/models/blob/master/official/nlp/configs/pretraining_experiments.py#L25
We don't need to subclass.

File: official/projects/lra/transformer_encoder.py [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+
+from official.modeling import tf_utils
+from official.nlp.modeling import layers
+
+_Initializer = Union[str, tf.keras.initializers.Initializer]
+_approx_gelu = lambda x: tf.keras.activations.gelu(x, approximate=True)
+
+
+class TransformerEncoder(tf.keras.layers.Layer):
+  """TransformerEncoder.
```
Comment (saberkun) [inline official/projects/lra/transformer_encoder.py:31]: I feel we need docstring to explain what's the difference of this guy v.s. the bert?

**Review by saberkun — CHANGES_REQUESTED** [review]

File: official/projects/lra/transformer_encoder.py (hunk 9 above)
Comment (EvanZhuang) [reply official/projects/lra/transformer_encoder.py:31]: They are mostly similar, with small adaptations for the pipeline. If you think we need a docstring I'll update that.

File: official/projects/lra/transformer_experiments.py (hunk 8 above)
Comment (EvanZhuang) [reply official/projects/lra/transformer_experiments.py]: Removed

File: official/projects/lra/linformer_encoder_block.py (hunk 7 above)
Comment (EvanZhuang) [reply official/projects/lra/linformer_encoder_block.py:215]: Yes, it has an additional dimension reduction at the front

File: official/projects/lra/linformer_encoder_block.py (hunk 6 above)
Comment (EvanZhuang) [reply official/projects/lra/linformer_encoder_block.py]: Removed, yes

File: official/projects/lra/linformer_encoder_block.py (hunk 5 above)
Comment (EvanZhuang) [reply official/projects/lra/linformer_encoder_block.py]: Removed

File: official/projects/lra/lra_dual_encoder_task.py (hunk 4 above)
Comment (EvanZhuang) [reply official/projects/lra/lra_dual_encoder_task.py:1]: Could we add a unit test to the lra dual encoder task? -> Sure, do you have a reference for this? I'll know better what to do

I am wondering if we can subclass the nlp/tasks/dual_encoder.py? -> Probably can't, I adapted from that code but these two are very different

File: official/projects/lra/lra_dual_encoder.py (hunk 3 above)
Comment (EvanZhuang) [reply official/projects/lra/lra_dual_encoder.py]: Removed

File: official/projects/lra/lra_dual_encoder.py (hunk 2 above)
Comment (EvanZhuang) [reply official/projects/lra/lra_dual_encoder.py]: Removed

File: official/projects/lra/linformer_experiments.py (hunk 1 above)
Comment (EvanZhuang) [reply official/projects/lra/linformer_experiments.py]: Updated this script and others accordingly

File: official/projects/lra/linformer_encoder.py [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+    self.inputs = dict(input_word_ids=tf.keras.Input(shape=(None,),
+                                                     dtype=tf.int32),
+                       input_mask=tf.keras.Input(shape=(None,), dtype=tf.int32),
+                       input_type_ids=tf.keras.Input(shape=(None,),
+                                                     dtype=tf.int32))
+
+  def call(self, inputs):
+    word_embeddings = None
+
+    if isinstance(inputs, dict):
```
Comment (CrickWu) [inline official/projects/lra/linformer_encoder.py:188]: Can we refactor this out?

File: official/projects/lra/linformer_encoder.py [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+  def get_embedding_table(self):
+    return self._embedding_layer.embeddings
+
+  def get_embedding_layer(self):
+    return self._embedding_layer
+
+  def get_config(self):
+    return dict(self._config)
+
+  @property
```
Comment (CrickWu) [inline official/projects/lra/linformer_encoder.py:255]: Personal opinion, we may not need these properties.

File: official/projects/lra/lra_dual_encoder.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+               network,
+               num_classes,
+               max_seq_length,
+               dropout_rate=0.1,
+               initializer='glorot_uniform',
+               use_encoder_pooler=True,
+               head_name='dual_encode',
+               **kwargs) -> None:
+
+    left_word_ids = tf.keras.layers.Input(shape=(max_seq_length,),
```
Comment (CrickWu) [inline official/projects/lra/lra_dual_encoder.py:54]: Can we make it a subclass Keras layer: e.g.. using `call()` to have these logics.

**Review by CrickWu — COMMENTED** [review]
Some nit comments.

File: official/projects/lra/linformer_encoder.py (hunk 10 above)
Comment (EvanZhuang) [reply official/projects/lra/linformer_encoder.py:188]: refactored! let me know if you meant something else

File: official/projects/lra/linformer_encoder.py (hunk 11 above)
Comment (EvanZhuang) [reply official/projects/lra/linformer_encoder.py:255]: properties removed, still works after a simple test

File: official/projects/lra/lra_dual_encoder.py (hunk 12 above)
Comment (EvanZhuang) [reply official/projects/lra/lra_dual_encoder.py:54]: This part was following `official/nlp/modeling/models/dual_encoder.py` do you have an example for the keras layer style? Be happy to adapt to that

Comment (saberkun) [reply official/projects/lra/lra_dual_encoder.py:54]: Hi Evan, you can take a look at: https://github.com/tensorflow/models/blob/master/official/nlp/modeling/networks/bert_encoder.py#L223
The BertEncoderV2 is modified to be a keras layer and using call() override.

File: official/projects/lra/lra_dual_encoder_task.py (hunk 4 above)
Comment (saberkun) [reply official/projects/lra/lra_dual_encoder_task.py:1]: No problem

File: official/projects/lra/lra_dual_encoder.py (hunk 12 above)
Comment (EvanZhuang) [reply official/projects/lra/lra_dual_encoder.py]: Keras layer and call() rewrite finished

**Review by saberkun — APPROVED** [review]
